warehouse/forklift/legacy.py

```diff
@@ -1161,6 +1161,7 @@ def file_upload(request):
                 else "OpenID created token",
                 "canonical_version": release.canonical_version,
                 "publisher_url": request.oidc_publisher.publisher_url
+                + f"/commit/{request.oidc_claims.get('sha', 'unknown')}"
```

Comment:
We should degrade more gracefully than this: if `sha` isn't present, then let's just render the bare `publisher_url` without any `/commit/...` path.

I'd be okay with abstracting that behind some kind of helper on `OIDCPublisher`, e.g.:

```
"publisher_url": request.oidc_publisher.publisher_url(request.oidc_claims)
```

In other words: `publisher_url` could go from being a `@property` to a method that progressively enhances based on the state in `request.oidc_claims`.